File: _programmers/_02_/truck.java
// import java.util.Arrays;
// import java.util.List;
import java.util.Queue;
// import java.util.function.Supplier;
import java.util.LinkedList;
// import java.util.ArrayList;
// import java.util.stream.Collectors;
// import java.util.stream.IntStream;

class Solution {
  public int solution(int bridge_length, int weight, int[] truck_weights) {
    // ArrayList<Integer> list = (ArrayList<Integer>) IntStream
    //     .of(truck_weights)
    //     .boxed()
    //     .collect(Collectors.toCollection(ArrayList::new));

    // ArrayList<Integer> _list = Arrays
    //   .stream(truck_weights)
    //   .boxed()
    //   .collect(Collectors.toCollection(ArrayList::new));
  
    // 위의 두 가지 변환 방법 모두 컴파일러가 정확한 타입 정보를 추론하지 못함
    // truck_weights의 타입 정보가 stream 처리시 반영되지 않음
    // 자료형 선언/할당을 하지 않으면 타입 에러가 발생함(or 타입 캐스팅 필요함)

    // List<Integer> list = Arrays
    //   .stream(truck_weights)
    //   .boxed()
    //   .collect(Collectors.toList());

    // List<Integer> list = Arrays.stream(truck_weights)
    // .boxed()
    // .collect(Collectors.toList());

    // https://www.nextree.co.kr/p6506/
    // ArrayList 연속된 주소(배열과 유사)
    // LinkedList 포인터로 연결

    // Supplier<List<Integer>> supplier = () -> new LinkedList<Integer>();
    // Queue<Integer> truck_q = (LinkedList<Integer>) Arrays.stream(truck_weights)
    //   .boxed()
    //   .collect(Collectors.toCollection(supplier));
    

    int time = 0;
    Queue<Integer> trucks_on_brige = new LinkedList<>();

    int weightSum = 0;
    for (int w : truck_weights) {
      // System.out.println(String.format("truck: %s", w));
      while(true) {
        // producing
        if (trucks_on_brige.size() < bridge_length) {
          if (weightSum + w > weight) {
            trucks_on_brige.offer(0);
            // time++;
            // System.out.println(String.format("over loaded, weightSum + w: %s / time: %s / trucks on bridge: %s", weightSum + w, time, trucks_on_brige));
          } else {
            trucks_on_brige.offer(w);
            weightSum += w;
            // time++;
            // System.out.println(String.format("add truck, weightSum: %s / time: %s / trucks on bridge: %s", weightSum, time, trucks_on_brige));
            break;
          }
        // consuming
        } else {
          weightSum -= trucks_on_brige.poll();
          time++;
          // System.out.println(String.format("comsuming, weightSum: %s / time: %s / trucks on bridge: %s", weightSum, time, trucks_on_brige));
        }
      }
    }
    return bridge_length + time + bridge_length;
  }
}

class Truck {
  public static void main(String[] args) {
    System.out.println(
    new Solution()
      .solution(
        2, 
        10, 
        new int[]{ 7, 4, 5, 6 }
        )
    );
  }
}

File: _programmers/_02_/truck.kt
// https://programmers.co.kr/learn/courses/30/lessons/42583

import java.util.*

// 다리의 길이, 최대 중량, 트럭 중량 리스트가 주어짐
// 시간을 담을 변수를 0으로 초기화함
// 먼저 다리를 건너는 트럭을 표현하기 위해 다리의 길이 크기의 큐를 생성하고 0으로 초기화함
// 다음으로 트럭 중량이 담을 큐를 생성하고 리스트를 순회하며 관련 요소를 옮김
// 다리를 건너는 트럭 큐가 소진될때까지 recursive loop로 순회를 반복함
// 루프 내에서 시간 카운터가 하나 증가하며, 다리를 건너는 트럭 큐에서 하나의 요소를 polling함
// 대기열의 트럭 중 첫번째 요소와 다리를 건너는 트럭의 총 중량을 합한 값을 총중량과 비교
// 만일 값이 작으면 대기열의 트럭에서 요소를 하나 뽑아서 다리를 건너는 트럭 큐에 추가함
// 만일 값이 크면 대기열의 트럭에서 요소를 뽑는 댓딘 다리를 지나는 트럭 큐에 0을 추가함
// 모든 요소를 순회한 이후에는 경과된 시간 카운터를 반환함

class Solution02 {
  fun solution(bridge_length: Int, weight: Int, truck_weights: IntArray): Int {
      var time = 0
      var trucksOnBridge: Queue<Int> = LinkedList()
      var truckQ: Queue<Int> = LinkedList()

      for (i in 0 until bridge_length) {
        trucksOnBridge.add(0)
      }

      for (i in 0..truck_weights.lastIndex) {
        truckQ.offer(truck_weights[i])
      }

      while (!trucksOnBridge.isEmpty()) {
        trucksOnBridge.poll()
        if (!truckQ.isEmpty()) {
          val totalWeight = trucksOnBridge.fold<Int, Int>(0) { 
            a, b -> a + b 
          }
          if (totalWeight + truckQ.peek() <= weight) {
            val truck = truckQ.poll()
            trucksOnBridge.add(truck)
          } else {
            trucksOnBridge.add(0)
          }
          time++
        }
      }
      return time
  }
}

fun main(args: Array<String>) {
    Solution02()
      .solution(2, 10, intArrayOf(7, 4, 5, 6))
}